aten/src/ATen/native/cuda/UpSample.cuh

```diff
@@ -197,6 +197,15 @@ __device__ __forceinline__ static int nearest_neighbor_compute_source_index(
   return src_index;
 }
 
+__device__ __forceinline__ static int nearest_neighbor_bw_compute_source_index(
+    const float scale,
+    int dst_index,
+    int output_size) {
+  const int src_index =
```

Comment:
Could you explain why we shouldn't re-use the same indices as in forward here?
Why is this wrong? "The value used during the forward should be the one that gets the gradients."

Reply:
>  "The value used during the forward should be the one that gets the gradients."
This is exactly what is desired.

The tricky thing is, we are lauching kernel on different configuration for fw & bw. Forward was launched with threads mapping to output, while backward was launched with threads mapping to input (so we could avoid atomicAdd, which is slow for fp16).

Hence on the backward, the mapping is opposite from forward, so we can't use the same function.

Reply:
Funny story, in my initial implementation, I actually got it right: https://github.com/pytorch/pytorch/pull/21694/commits/3284764ec657c86d7c1a067d930538bede3f3245#diff-5092da792c30694ee4adf0d0ae2a37c6R166
But as I continue tuning the implmentation in the PR, I somehow got confused by this myself and thought maybe the same mapping could be reused. Unfortunately CI didn't catch my mistake and here we are. :man_facepalming:

Reply:
Ok!

I think the main confusing bit for me is that the kernel is called as `upsample_nearest1d_backward_out_frame(odata, nbatch, channels, output_width, input_width, idata, scale_factor);` while the arguments in the function implementation are `upsample_nearest1d_backward_out_frame(
    const scalar_t* grad_o, size_t dim_b, size_t dim_c, size_t src_dim_w, size_t dst_dim_w, scalar_t* grad_i, float scale_factor)`. So `output_width` is given as `src_dim_w` and `input_width` becomes `dst_dim_w`.
So `dst_idx` indexes inside the output in the forward code and inside the input (grad_input) in the backward code!
And the same "scale" during forward is "1/scale" during backward!